Firestore/Source/Util/FSTDispatchQueue.mm

```diff
@@ -56,9 +187,70 @@ - (void)dispatchAsyncAllowingSameQueue:(void (^)(void))block {
   dispatch_async(self.queue, block);
 }
 
-- (void)dispatchAfterDelay:(NSTimeInterval)delay block:(void (^)(void))block {
-  dispatch_time_t delayNs = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delay * NSEC_PER_SEC));
-  dispatch_after(delayNs, self.queue, block);
+- (FSTDelayedCallback *)dispatchAfterDelay:(NSTimeInterval)delay
+                                   timerID:(FSTTimerID)timerID
+                                     block:(void (^)(void))block {
+  // While not necessarily harmful, we currently don't expect to have multiple callbacks with the
+  // same timerID in the queue, so defensively reject them.
+  FSTAssert(![self containsDelayedCallbackWithTimerID:timerID],
+            @"Attempted to schedule multiple callbacks with id %ld", (unsigned long)timerID);
+  FSTDelayedCallback *delayedCallback = [FSTDelayedCallback createAndScheduleWithQueue:self
+                                                                               timerID:timerID
+                                                                                 delay:delay
+                                                                              callback:block];
+  [self.delayedCallbacks addObject:delayedCallback];
+  return delayedCallback;
+}
+
+- (BOOL)containsDelayedCallbackWithTimerID:(FSTTimerID)timerID {
+  NSUInteger matchIndex = [self.delayedCallbacks
+      indexOfObjectPassingTest:^BOOL(FSTDelayedCallback *obj, NSUInteger idx, BOOL *stop) {
+        return obj.timerID == timerID;
+      }];
+  return matchIndex != NSNotFound;
+}
+
+- (void)runDelayedCallbacksUntil:(FSTTimerID)lastTimerID {
+  dispatch_semaphore_t doneSemaphore = dispatch_semaphore_create(0);
+
+  [self dispatchAsync:^{
+    FSTAssert(lastTimerID == FSTTimerIDAll || [self containsDelayedCallbackWithTimerID:lastTimerID],
+              @"Attempted to run callbacks until missing timer ID: %ld",
+              (unsigned long)lastTimerID);
+
+    [self sortDelayedCallbacks];
+    for (FSTDelayedCallback *callback in self.delayedCallbacks) {
+      [callback skipDelay];
+      if (lastTimerID != FSTTimerIDAll && callback.timerID == lastTimerID) {
+        break;
+      }
+    }
+
+    // Now that the callbacks are queued, we want to enqueue an additional item to release the
+    // 'done' semaphore.
+    [self dispatchAsyncAllowingSameQueue:^{
+      dispatch_semaphore_signal(doneSemaphore);
+    }];
+  }];
+
+  dispatch_semaphore_wait(doneSemaphore, DISPATCH_TIME_FOREVER);
+}
+
+- (void)sortDelayedCallbacks {
```

Comment:
This is an O(n^2 ln(n)) implementation of a priority queue (n sorts for n events). You could make this O(n lg(n)) if you used `std::priority_queue` as the backing store. See how greg did it here:

https://github.com/firebase/firebase-ios-sdk/pull/695/files#diff-52dc866a3432d82fec4b0b487d8a41b0R43

Reply:
This is only called by tests and is actually "constant time" because there are only 4 timer IDs, and we disallow duplicates right now, so the list is no bigger than 4 elements. :-) I added a comment.